### PR TITLE
exp/ingest: Add Ingestion primitives

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Ignore the commits before https://github.com/stellar/go/commit/cb22fd97158660bd19274770503f1aa4c7327566 since they are part of other PRs.

This PR adds functions to run processors on change / transaction streams.

### Why

Fixes https://github.com/stellar/go/issues/2169

>We should provide some library functions in exp/ingest package to interact with ChangeProcessor and LedgerTransactionProcessor instances. These functions should address the following use cases which are required for implementing horizon's ingestion system:

 >Applying a ChangeProcessor on all the entries of a history archive snapshot
 Applying a ChangeProcessor on all the changes from a single ledger
 Applying a TransactionProcessor on all the transactions from a single ledger

### Known limitations

* We should use the context in the state reader and ledger / change reader to abort early if the context has already terminated. This functionality can be added in another commit if there is agreement on this behavior

### CODE REVIEW

File: exp/ingest/io/change_reader.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+// ChangeReader provides convenient, streaming access to changes within a ledger.
+type ChangeReader interface {
+	GetSequence() uint32
+	GetHeader() xdr.LedgerHeaderHistoryEntry
+	// Read should return the next `Change` in the leader. If there are no more
+	// changes left it should return an `io.EOF` error.
+	Read() (Change, error)
+}
```
Comment (bartekn) [inline exp/ingest/io/change_reader.go]: > We should use the context in the state reader and ledger / change reader to abort early if the context has already terminated.

Why not use `Close()` method? Can be used easily with `defer`.

File: exp/ingest/processors.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	processors ...LedgerTransactionProcessor,
+) LedgerTransactionProcessor {
+	return ltGroup(processors)
+}
+
+func ProcessLedgerTransactions(
+	ctx context.Context,
+	txProcessor LedgerTransactionProcessor,
+	ledgerBackend ledgerbackend.LedgerBackend,
+	ledger uint32,
```
Comment (bartekn) [inline exp/ingest/processors.go]: Maybe instead of `ledgerBackend` and `ledger` just pass `LedgerReader`?

File: exp/ingest/processors.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+				err,
+				"could not process transaction %v in ledger %v",
+				tx.Index,
+				ledger,
+			)
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```
Comment (bartekn) [inline exp/ingest/processors.go]: We should probably copy `select` to `Init` and `Commit` stages.

File: exp/ingest/processors.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	if err = changeProcessor.Commit(); err != nil {
+		return errors.Wrapf(err, "could not commit processor for sequence %v", ledger)
+	}
+
+	return nil
+}
+
+func ProcessLedgerChanges(
```
Comment (bartekn) [inline exp/ingest/processors.go]: I wonder if we should merge `ProcessLedgerChanges` and `ProcessHAS`. Internally, they are almost the same.  Maybe we can have a single `ProcessChanges(ctx context.Context, changeProcessor ChangeProcessor, reader ChangeReader)` function and add an adapter that wraps `StateReader` to change `xdr.LedgerEntryChange` to `io.Change` (super easy). OR change `StateReader` to stream `io.Change`.

**Comment by tamirms** [comment]
@bartekn can you take another look at this PR? It should implement all the changes we discussed earlier today.

The interfaces are pretty minimal but it should be sufficient for progressing with the horizon ingestion refactor. We should revisit the interfaces and potentially add more functions to address common use cases before releasing the ingestion package to external developers

File: exp/ingest/io/processors.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+type ChangeProcessor interface {
+	ProcessChange(change Change) error
+}
+
+type LedgerTransactionProcessor interface {
+	ProcessTransaction(transaction LedgerTransaction) error
+}
+
+func StreamLedgerTransactions(
```
Comment (bartekn) [inline exp/ingest/io/processors.go]: I wonder if `Stream` explains what happens inside properly.
```suggestion
func ProcessLedgerTransactions(
```

File: exp/ingest/io/processors.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+			return errors.Wrapf(
+				err,
+				"could not process transaction %v",
+				tx.Index,
+			)
+		}
+	}
+}
+
+func StreamChanges(
```
Comment (bartekn) [inline exp/ingest/io/processors.go]: I wonder if `Stream` explains what happens inside properly.
```suggestion
func ProcessChanges(
```

File: exp/tools/horizon-demo/orderbook_processor.go [hunk 7]
```diff
@@ -48,7 +48,6 @@ func (p *OrderbookProcessor) ProcessState(ctx context.Context, store *pipeline.S
 }
 
 func (p *OrderbookProcessor) ProcessLedger(ctx context.Context, store *pipeline.Store, r io.LedgerReader, w io.LedgerWriter) error {
-	defer r.Close()
```
Comment (bartekn) [inline exp/tools/horizon-demo/orderbook_processor.go]: Added https://github.com/stellar/go/issues/2180.

File: exp/ingest/io/change_reader.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+// ChangeReader provides convenient, streaming access to a sequence of Changes
+type ChangeReader interface {
+	// Read should return the next `Change` in the leader. If there are no more
+	// changes left it should return an `io.EOF` error.
+	Read() (Change, error)
+}
+
+// LedgerChangeReader is a ChangeReader which returns Changes from Stellar Core
+// for a single ledger
+type LedgerChangeReader struct {
```
Comment (bartekn) [inline exp/ingest/io/change_reader.go]: We need tests.

**Comment by tamirms** [comment]
most of the code from this PR is in https://github.com/stellar/go/pull/2178 so we can close this PR

**Pull request closed without merging**